### PR TITLE
Authentication Integration

### PR DESCRIPTION
**Pull request by ugur-akin**
### Implemented* authentication logic on front-end:
- Implemented a User api to send ajax requests to server.
- Implemented `useEffect` hooks to call User api methods:
  - App component calls `ResolveSession` upon initial load.
  - Signup and Login form components call `Create `and `Authenticate` respectively upon submission.
- Implemented wrapped routes:
  - `PrivateRoute`s redirect to login if user is not authenticated.
  - `AuthRoute`s redirect to Home/Dashboard if the user is already authenticated.
- Implemented a "first-draft" `UserContext`.

### Made some changes to the backend: 
- Loosely standardized error responses and modified error status codes.
- **IMPORTANT**: Modified signup and login end-point requirements: 
  - Upon signup (now `user/create`), server ~~no longer logs user in~~ or returns ~~user~~/jwt in the body.
  - Upon login (now `user/authenticate`), server no longer returns jwt in the body. Instead, jwt is sent via an `httpOnly` cookie.
- Modified `auth` middleware to look for a jwt sent within cookies.
- Added session related end-points to user routes.
  - `user/session/resolve`: Resolves session by verifying the jwt sent via authentication cookie. Responds with user object upon success.
  - `user/session/extend` (Partially implemented): Issues a new jwt to extend expiry date. Invalidates the old jwt.
  - ` user/session/end` (Not implemented): Resets the authentication cookie and invalidates the jwt sent by authentication cookie.
- Implemented some unit tests.

Edit: User is now logged in upon successful registration.

<sub>*: Frontend/Backend api is fully functional for current needs, however, implementation isn't complete (e.g. no way to end session/sign out). Also some refactoring and clean-up is needed.</sub>

### CODE REVIEW

File: server/routes/user.js [hunk 1]
```diff
@@ -8,102 +8,137 @@ const jwt = require("jsonwebtoken");
...
+
+    // Bcrypt password
+    const salt = await bcrypt.genSalt(10);
+    user.password = await bcrypt.hash(password, salt);
+    await user.save();
+
+    res.status(201).json();
+  } catch (err) {
+    console.error(err.message);
+    res.status(500).send("Server Error");
```
Comment (ThinkTink) [inline server/routes/user.js]: This should be updated to match the other routes ({ message:...

File: client/src/api/User.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  // TODO: Implement error handling
+
+  create: async (data) => {
+    try {
+      await axios.post('/user/create', data);
+      const success = true;
+      return {success};
+    } catch (err) {
+      const success = false;
+      const {message} = err.body;
```
Comment (ThinkTink) [inline client/src/api/User.js]: Should this be "err.response.data" like the other routes?

File: client/src/components/AuthRoute.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import {Route, Redirect} from 'react-router-dom';
+
+const AuthRoute = ({component: Component, authed, ...rest}) => {
+  return (
+    <Route
+      {...rest}
+      render={(props) =>
+        authed === false ? (
```
Comment (ThinkTink) [inline client/src/components/AuthRoute.js]: This could be simplified to " !authed ? ( "

File: client/src/components/PrivateRoute.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import {Route, Redirect} from 'react-router-dom';
+
+const PrivateRoute = ({component: Component, authed, ...rest}) => {
+  return (
+    <Route
+      {...rest}
+      render={(props) =>
+        authed === true ? (
```
Comment (ThinkTink) [inline client/src/components/PrivateRoute.js]: and this could be simplified to " authed ? ( "

File: server/routes/user.js [hunk 5]
```diff
@@ -8,102 +8,137 @@ const jwt = require("jsonwebtoken");
...
+
+  try {
+    // User exists
+    let isUser = await User.findOne({ email });
+    if (isUser) {
+      return res.status(403).send({ message: "Email is already in use!" });
+    }
+
+    // Create new user
+    let user = new User({
```
Comment (ThinkTink) [inline server/routes/user.js]: isUser and user can be const rather than let

File: server/routes/user.js [hunk 6]
```diff
@@ -8,102 +8,137 @@ const jwt = require("jsonwebtoken");
...
-
+    const { email, password } = req.body;
     try {
       // user exists
-      let isUser = await User.findOne({ email });
-      if (isUser) {
-        return res
-          .status(400)
-          .send({ errors: [{ msg: "User already exists" }] });
+      let user = await User.findOne({ email });
```
Comment (ThinkTink) [inline server/routes/user.js]: this can be const as well

File: server/middleware/auth.js [hunk 7]
```diff
@@ -1,20 +1,23 @@
 const jwt = require("jsonwebtoken");
 const keys = require("../config/keys");
 
+//  TODO: Refactor cookie name "auth-token" into
+//  a variable? (also used in user/authenticate route)
 module.exports = function (req, res, next) {
-  //get the token from the header
-  const token = req.header("x-auth-token");
+  // Get token from http-only cookie
```
Comment (ThinkTink) [inline server/middleware/auth.js]: To make this consistent with validator, perhaps it should be named authenticator rather than auth?

**Comment by ThinkTink** [comment]
Good work on this! I pointed out some small things to clean up

**Comment by ugur-akin** [comment]
> Good work on this! I pointed out some small things to clean up

Thanks for the feedback, will address those issues next.

**Pull request merged**